tsdb/meta.go

```diff
@@ -760,18 +761,18 @@ func (m *Measurement) filters(condition influxql.Expr) (map[uint64]influxql.Expr
 // TODO: this shouldn't be exported. However, until tx.go and the engine get refactored into tsdb, we need it.
 func (m *Measurement) TagSets(dimensions []string, condition influxql.Expr) ([]*influxql.TagSet, error) {
 	m.mu.RLock()
-	defer m.mu.RUnlock()
 
 	// get the unique set of series ids and the filters that should be applied to each
 	filters, err := m.filters(condition)
 	if err != nil {
+		m.mu.RUnlock()
 		return nil, err
 	}
 
 	// For every series, get the tag values for the requested tag keys i.e. dimensions. This is the
 	// TagSet for that series. Series with the same TagSet are then grouped together, because for the
 	// purpose of GROUP BY they are part of the same composite series.
-	tagSets := make(map[string]*influxql.TagSet)
+	tagSets := make(map[string]*influxql.TagSet, len(filters))
```

Comment:
I agree. This certainly needs to change in some way and the `AddFilter` below also needs to. I'm just not sure the proper way to do it.